src/rmind/callbacks/loggers.py

```diff
@@ -106,3 +112,189 @@ def _call(
+        IMAGE = "image"
+        WAYPOINTS_XY_NORMALIZED = "waypoints_xy_normalized"
+        WAYPOINTS_XY = "waypoints_xy"
+        EGO_XY = "ego_xy"
+
+    def __init__(
+        self,
+        *,
+        select: dict[str, str | tuple[str, ...]],
+        when: Annotated[str, AfterValidator(_validate_hook)],
+        key: str,
+        every_n_batch: int | None = None,
+        crs: str | None = None,
+    ) -> None:
+        self._key = key
+        if every_n_batch is not None and when not in BATCH_HOOKS:
+            msg = (
+                "`every_n_batch` is only supported for batch-based hooks: "
+                + ", ".join(f"`{hook}`" for hook in BATCH_HOOKS)
+                + f". Got `{when}`"
+            )
+            raise ValueError(msg)
+        self._every_n_batch = every_n_batch
+        self._when = when
+        self._select = {k: tuple(v) for k, v in select.items()}
+        self._crs = crs
+        setattr(self, when, self._call)
+
+    def _call(
+        self,
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        if trainer.sanity_checking or not (
+            loggers := [
+                logger
+                for logger in pl_module.loggers
+                if isinstance(logger, WandbLogger)
+            ]
+        ):
+            return
+
+        base_hook_method = getattr(pl.Callback, self._when)
+        sig = inspect.signature(base_hook_method)
+
+        bound_args = sig.bind(self, trainer, pl_module, *args, **kwargs)
+        bound_args.apply_defaults()
+        batch_idx = bound_args.arguments.get("batch_idx")
+        batch = bound_args.arguments.get("batch")
+
+        if (
+            (self._every_n_batch is not None)
+            and (batch_idx is not None)
+            and (batch_idx % self._every_n_batch != 0)
+        ) or (trainer.current_epoch != 0):
+            return
+
+        batch = TensorDict(batch).auto_batch_size_()[self.in_batch_idx]
+        caption = []
+        if self.Columns.INPUT_ID in self._select:
+            caption.append(
+                batch.get(self._select[self.Columns.INPUT_ID])
+                .data[self.in_batch_idx]
+                .item()
+            )
+        if self.Columns.TIME_STAMP in self._select:
+            time_stamp = (
+                batch.get_at(self._select[self.Columns.TIME_STAMP], self.in_clip_idx)
+                .cpu()
+                .item()
+            )
+            caption.append(
+                datetime.fromtimestamp(time_stamp * 1e-6, tz=UTC).strftime(
+                    "%Y-%m-%d %H:%M:%S"
+                )
+            )
+        if self.Columns.FRAME_IDX in self._select:
+            frame_idx = (
+                batch.get_at(self._select[self.Columns.FRAME_IDX], self.in_clip_idx)
+                .cpu()
+                .item()
+            )
+            caption.append(f"frame_idx: {frame_idx}")
+        caption = " | ".join(caption)
```

Comment:
~~we have tensordict at home~~

i have a hunch this can be simplified by
- leaving batch as is
- specifying pytree-compatible paths in config (e.g. `['meta', 'input_id', 0]`) and dropping the hardcoded `self.in_batch_idx` / `self.in_clip_idx`"
- using `key_get(batch, path)`